### PR TITLE
Implemented Empty State for Course Instances

### LINKED ISSUE #10963: Add empty state for course instances page
The course instances page within a course should have a meaningful empty state that briefly describes what a course instance is, links to the documentation for more information, and provides a clear call to action to create a new course instance.

The page is implemented in [`apps/prairielearn/src/pages/instructorCourseAdminInstances`](https://github.com/PrairieLearn/PrairieLearn/tree/master/apps/prairielearn/src/pages/instructorCourseAdminInstances).

### PR DESCRIPTION
**Pull request by miguelaenlle**
Closes #10963.

On the `instructorCourseAdminInstances` page, if a course has no course instances, the following empty state is displayed:

Screenshot 2024-12-09 at 8 31 59 PM

### CODE REVIEW

**Comment by nwalters512** [comment]
@miguelaenlle no need to accept the CLA - I'll add you to our allowlist since you're an intern! I'll do so tomorrow.

**Comment by miguelaenlle** [comment]
I have reviewed and hereby sign the CLA

**Comment by miguelaenlle** [comment]
> @miguelaenlle no need to accept the CLA - I'll add you to our allowlist since you're an intern! I'll do so tomorrow.

Sounds great! Thank you @nwalters512.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:133]: I think we should use something other than `<code>` for referring to buttons, because it uses a monospaced font that's different to the actual buttons. Elsewhere we've used `<strong>` to refer to buttons, but this is just an idea.

Comment (miguelaenlle) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: Hi Matt. I agree, and made this change:

image

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:124]: What do you think about conditionally rendering the whole table? That is, if there aren't any course instances, it may make sense to hide the table and its headers, as the headers don't do anything in the empty state.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:126]: At specific viewports, the text comes really close to the side of the panel:

Screenshot 2024-12-10 at 09 16 25

I'd recommend using the `card-body` class here to provide appropriate padding: https://getbootstrap.com/docs/5.3/components/card/#body

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:131]: The text here can appear unbalanced at specific viewports:

Screenshot 2024-12-10 at 09 17 47

There's a newish CSS feature called `text-wrap: balance` that could improve this: https://developer.mozilla.org/en-US/docs/Web/CSS/text-wrap#balance. What do you think about adding that in an inline `style` attribute here?

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:134]: The "Add course button" is actually conditionally rendered, see lines 35-37:

- It isn't shown for the example course.
- It isn't shown if the user doesn't have Editor permissions for the course.
- It isn't shown if the course path doesn't exist on disk (this is the `resLocals.needToSync` flag).

We often find that people get confused about why they can't do something, so it'd improve the UX to show specific messages in these cases. For instance, if `resLocals.authz_data.has_course_permission_edit` is `false`, we can say something like "Course Editors can create new course instances". How does that sound?

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:138]: There are two pages that discuss course instances: the one you linked to, and https://prairielearn.readthedocs.io/en/latest/courseInstance/. Any preference which one we should link to? cc @mfsilva22, who probably has thoughts on links from PL to the documentation.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:130]: cc @mfsilva22 to review the copy here.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:133]: It might be nice to show an inline call to action here. For instance, that's what the [GitHub wiki pages](https://github.com/PrairieLearn/cla/wiki) does if there aren't any pages yet:

Screenshot 2024-12-10 at 09 28 49

**Review by nwalters512 — COMMENTED** [review]
🎉 congrats on your first PR! I left some comments to drive discussion. I know there are a lot of comments even though there are a small number of lines changed, don't worry! This is our chance to figure out patterns that we'll use across the other pages.

**Comment by miguelaenlle** [comment]
I have reviewed and hereby sign the CLA

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (miguelaenlle) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: That sounds good -- I modified the page to hide the table headers if there are no course instances:

image

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (miguelaenlle) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: I added the card-body padding:

image

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (miguelaenlle) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: I added the text-wrap: balance styling, so this should be fixed now.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (miguelaenlle) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: The UI now displays messages in the cases where the Add course instance button is hidden:
- Doesn't have Editor permissions: "Course Editors can create new course instances."
- Example course: "You can't add course instances to the Example Course."
- Needs to Sync: "Sync the course from the Sync page before creating a new course instance."

Let me know if I should adjust any of these messages.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (miguelaenlle) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: I changed the link to this one -- this makes more sense, since the one I had before linked to instructions to create a Course Instance without the modal I'm developing in #11004.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (miguelaenlle) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: Agreed -- I added this:

image

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: Looking pretty good! I'll leave some comments on the specific messages; we can continue conversation there.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: Good thinking! Agreed that the new link is better.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: Looks great 💯

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:179]: This leaves us with the interesting possibility of showing multiple messages. I'd be inclined to show at most one, with the following precedence:

- Handle the example course first (permissions / syncing won't change if a course can be added)
- Handle the permissions next (one can't sync the course without editor permissions)
- Handle the "need to sync" case last

We have a tiny helper function (https://github.com/PrairieLearn/PrairieLearn/tree/master/packages/run) that can make this kind of thing nice to write, if you like it:

```ts
${run(() => {
  if (resLocals.course.example_course) {
    return html`<p>You can't add course instances to the Example Course.</p>`;
  }

  if (!resLocals.authz_data.has_course_permission_edit) {
    return html`<p>Course Editors can create new course instances.</p>`;
  }

  if (resLocals.needToSync) {
    return html`
      <p>
        Sync the course from the <strong>Sync</strong> page before creating a new course instance.
      </p>
    `;
  }

  return html`
    <form class="ml-auto" name="add-course-instance-form" method="POST">
      <input type="hidden" name="__csrf_token" value="${resLocals.__csrf_token}" />
      <button name="__action" value="add_course_instance" class="btn btn-sm btn-primary">
        <i class="fa fa-plus" aria-hidden="true"></i>
        <span>Add course instance</span>
      </button>
    </form>
  `;
})}
```

I opted to merge this with the code that determines when to show the button; this ensures that we always show either the button or exactly one message. Let me know if you like that.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:176]: It could be nice to link to the sync page here; something like "You must [sync the course](https://google.com) before creating a new course instance."

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:170]: Conventionally we refer to the example course in lower case ("example course").

**Review by nwalters512 — COMMENTED** [review]
Looking really good, here are a few more comments for consideration.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (miguelaenlle) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: Thanks for showing me the run function, @nwalters512 . I added this change.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (miguelaenlle) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: I added this change:

image

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (miguelaenlle) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: I added this change 👍

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:136]: Not yet requesting a change, just some food for thought when Mariana reviews: I'm wondering if it's better to try to describe what a course instance actually is rather than some example names. For instance, "A course instance contains the assessments and other configuration for a single offering of a course."

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:163]: Nit (`ml-auto` doesn't do anything here, and I don't think this form needs a name):

```suggestion
                        <form method="POST">
```

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:157]: Nit:

```suggestion
                            <a href="${resLocals.urlPrefix}/course_admin/syncs">sync this course</a>
```

**Review by nwalters512 — COMMENTED** [review]
A few final nits, nothing major! I'll wait for Mariana to review before I issue a final approval and merge.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (mfsilva22) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:136]: I like the suggestion by @nwalters512 above, but I think we should still offer an example (works as "best practice" in this case). I would just use one ("Fall 2024"). 

Based on my experience running queries, sometimes instructors use course instance names that are not helpful. Some include the name of the course, not knowing that the course name and course instance name get both appended when displayed in the "Courses" list

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (mfsilva22) [inline apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:144]: Based on my conversation with @mwest1066 about this project to improvement the onboarding experience, I understood the first goal was to remove the need to edit json files, or further, to remove the need from the user to understand the format of these json files. If that is the case, I don't think we should link to the documentation that explains infoCourseInstance.json. 

Is this just a placeholder for future better documentation that we will provide to users?

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:136]: I think we can suggest best practices in #11004 so that the suggestion is visible directly next to the input where they'll actually pick a name. Does that sound reasonable?

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:144]: Yep, I'm thinking that anything we put here is going to be a placeholder. Before we write documentation that talks about course instance config independently of JSON, we need to have creation/edit flows in the PL web UI. I figure at this point, some docs are better than none, and understanding how to edit JSON _will_ still be necessary until nice web editing UIs are in place.

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (mfsilva22) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:136]: yes, I Iike that!

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (mfsilva22) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts:144]: agree!

File: apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts]: Looks great!

**Review by nwalters512 — APPROVED** [review]

**Pull request merged**
